### PR TITLE
Add a new e2e test for openfaas cloud

### LINKED ISSUE #403: Reliability monitoring - e2e testing of GitHub workflow
Reliability monitoring - e2e testing of GitHub workflow

## Context

Since moving the Communtity Cluster from GKE to DigitalOcean the reliability has dropped dramatically due to issues with the Cilium networking driver 1.4.0. A separate action will be required for this, but in the meantime it highlights a lack of e2e monitoring for the GitHub workflow end-to-end.

## Task

Create a Golang function that will:

- clone the repo (it will be a Go function)
- modify the handler's response to include a specific string such as the current date/time stamp
- do a git push (using a secret)
- call the function in a loop with a generous timeout until the response contains the new string
- call into a slack webhook if the string is incorrect or not 200 at the end

## Constraints

* The repo in question should be cloned with the go-github library
* This should be an OpenFaaS function
* Use /tmp/ as a writable area
* For write access use the GitHub App approach as per alexellis/derek and openfaas/openfaas-cloud

Suggestions and input is welcome

### PR DESCRIPTION
**Pull request by matipan**
Signed-off-by: Matias Pan <matias.pan26@gmail.com>

## Description
This PR adds a new function that performs an end-to-end test of openfaas cloud. Fixes #403.

#### How does it work?
This function tests if OpenFaaS Cloud is correctly listening to github push events, building and re-deploying a function. To validate this the function does the following:
* Modify the source file of a Go function to respond with the current timestamp
* Call that function in a loop every 15 second until the function responds with the timestamp we expected

If it responded ok we return `true`. `false` is returned whenever the test failed or something bad happened.

#### Configure the function
@alexellis In order to put this function on the OpenFaas Cloud Community Cluster we need to configure the following parameters:
* **gateway**: gateway to the cluster
* **image**: name of the image of this function
* **functionOwner** and **functionRepo**: owner and repo of the function that it's being used to do the test
* **goFilePath**: the path to the Go file of the function,
* **functionURL**: URL where the function will be deployed,
* **applicationID**: This function authenticates with github as an application, once the application was installed(it needs permissions to write the contents of the repository) use the ID here,
* **secrets**: create two secrets on your OpenFaaS Cloud installation, one with the secret key of the github app and another one with the private key. Basically we need to do the same thing that was done on derek.

We also need to create a repository with a plain Go function and make sure that OpenFaaS Cloud is installed on that repository.

## How Has This Been Tested?

I tested this using my own installation of OpenFaaS Cloud on a GKE cluster. Whenever the result of the test was ok it responds with true:
ofc-test

## How are existing users impacted? What migration steps/scripts do we need?
None

## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

### CODE REVIEW

**Comment by alexellis** [comment]
What's the best way to run this?

**Comment by matipan** [comment]
Create a GitHub app like for Derek with read/write permissions to the contents of the repo. Add the secrets to the cluster and deploy the following function:
```
provider:
   name: faas
   gateway: # your own OpenFaaS gateway
 functions:
   e2e-test:
     lang: golang-http
     handler: ./
     image: matipan/e2e-test:v1
     environment:
       functionOwner: # owner of the repository of the function being used to test
       functionRepo: # repository name for the function
       goFilePath: ./hello-world/handler.go # file-path for the handler file
       secret_path: /var/openfaas/secrets/
       functionURL: # change this to the URL of the function that this will call to validate that the new deployment is up
       application_id: # change this for the installation number of your github app
       exec_timeout: 360s
       read_timeout: 360s
       write_timeout: 360s
     secrets:
       - github-secret-key
       - github-private-key 
```
You can fork [this example function](https://github.com/matipan/hello-world-go) that I used to run the test.

You will also need to add the GitHub app "user handle" to the customers file, otherwise ofc will ignore the push that this function does.

**Comment by alexellis** [comment]
This looks really well done, thank you @matipan

File: e2e-test/stack.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    environment:
+      functionOwner: matipan # owner of the repository of the function being used to test
+      functionRepo: hello-world-go # repository name for the function
+      goFilePath: ./hello-world/handler.go # file-path for the handler file
+      secret_path: /var/openfaas/secrets/
+      functionURL: https://matipan.ofc.matiaspan.me/hello-world # change this to the URL of the function you'll use to test
+      application_id: 25945 # change this for the installation number of your github app
+      exec_timeout: 360s
+      read_timeout: 360s
+      write_timeout: 360s
```
Comment (Asa-aals5) [inline e2e-test/stack.yml]: > [](url)
```suggestion
      write_timeout : 360s

1. ### _****_

```

**Comment by alexellis** [comment]
Closing as inactive, if we still want to do this work, let's raise a new issue to discuss. Thanks for the contribution.

**Pull request closed without merging**